<a href="https://colab.research.google.com/github/krisnual/E-commerce-main/blob/main/Data_mining.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Lakukan pra-pemrosesan data pada dataset yang terletak di "/content/Online Retail.xlsx".

## Memuat data

### Subtask:
Muat data dari file Excel ke dalam DataFrame pandas.


**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame.



In [1]:
import pandas as pd

df = pd.read_excel('/content/Online Retail.xlsx')

**Reasoning**:
Display the first few rows of the loaded DataFrame to verify the data is loaded correctly.



In [2]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

## Mengeksplorasi data

### Subtask:
Tampilkan beberapa baris pertama data dan periksa informasi umum seperti tipe data, nilai yang hilang, dan statistik deskriptif.


**Reasoning**:
Display the first 5 rows of the dataframe, check the data types and non-null values, and display descriptive statistics for numerical columns.



In [3]:
display(df.head())
display(df.info())
display(df.describe())

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


None

Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.081158                            NaN      96.759853   

          CustomerID  
count  406829.000000  
mean    15287.690570  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.600303

## Membersihkan data

### Subtask:
Tangani nilai-nilai yang hilang dan hapus duplikat (jika ada).


**Reasoning**:
Check for missing values, fill missing values in 'Description', drop rows with missing 'CustomerID', check for duplicates, drop duplicates, and display info.



In [4]:
print("Missing values before handling:")
display(df.isnull().sum())

df['Description'].fillna('Unknown', inplace=True)
df.dropna(subset=['CustomerID'], inplace=True)

print("\nMissing values after handling:")
display(df.isnull().sum())

print("\nNumber of duplicate rows before handling:")
display(df.duplicated().sum())

df.drop_duplicates(inplace=True)

print("\nNumber of duplicate rows after handling:")
display(df.duplicated().sum())

print("\nDataFrame info after handling missing values and duplicates:")
df.info()

Missing values before handling:


InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


Missing values after handling:


/tmp/ipython-input-184992294.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Description'].fillna('Unknown', inplace=True)


InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


Number of duplicate rows before handling:


np.int64(5225)


Number of duplicate rows after handling:


np.int64(0)


DataFrame info after handling missing values and duplicates:
<class 'pandas.core.frame.DataFrame'>
Index: 401604 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    401604 non-null  object        
 1   StockCode    401604 non-null  object        
 2   Description  401604 non-null  object        
 3   Quantity     401604 non-null  int64         
 4   InvoiceDate  401604 non-null  datetime64[ns]
 5   UnitPrice    401604 non-null  float64       
 6   CustomerID   401604 non-null  float64       
 7   Country      401604 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.6+ MB


## Mengubah tipe data

### Subtask:
Konversi kolom-kolom dengan tipe data yang tidak sesuai (misalnya, mengonversi kolom tanggal ke tipe data datetime).


**Reasoning**:
Convert the `InvoiceDate` column to datetime objects and verify the data type.



In [5]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
print("Data type of InvoiceDate after conversion:")
display(df['InvoiceDate'].dtype)

Data type of InvoiceDate after conversion:


dtype('<M8[ns]')

## Menangani outlier

### Subtask:
Identifikasi dan tangani outlier (jika perlu).


**Reasoning**:
Identify numerical columns and calculate descriptive statistics to understand the distribution and potential outliers.



In [6]:
numerical_cols = ['Quantity', 'UnitPrice']
print("Descriptive statistics for numerical columns:")
display(df[numerical_cols].describe())

Descriptive statistics for numerical columns:


Quantity      UnitPrice
count  401604.000000  401604.000000
mean       12.183273       3.474064
std       250.283037      69.764035
min    -80995.000000       0.000000
25%         2.000000       1.250000
50%         5.000000       1.950000
75%        12.000000       3.750000
max     80995.000000   38970.000000

**Reasoning**:
Address negative and zero values in 'Quantity' and 'UnitPrice' as they represent non-standard transactions or errors, then handle extreme outliers using quantile-based filtering.



In [7]:
# Filter out rows with negative or zero Quantity and UnitPrice
df_filtered = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0)].copy()

# Calculate quantiles for outlier handling
q_quantity_high = df_filtered['Quantity'].quantile(0.99)
q_unitprice_high = df_filtered['UnitPrice'].quantile(0.99)

# Filter out extreme outliers based on quantiles
df_cleaned = df_filtered[(df_filtered['Quantity'] <= q_quantity_high) & (df_filtered['UnitPrice'] <= q_unitprice_high)].copy()

print("Descriptive statistics after handling outliers:")
display(df_cleaned[numerical_cols].describe())

Descriptive statistics after handling outliers:


Quantity      UnitPrice
count  385081.00000  385081.000000
mean       10.00891       2.731160
std        14.64235       2.545555
min         1.00000       0.001000
25%         2.00000       1.250000
50%         6.00000       1.950000
75%        12.00000       3.750000
max       120.00000      14.950000

## Summary:

### Data Analysis Key Findings

*   The initial dataset contained 541,909 entries with 8 columns.
*   Missing values were present in the 'Description' column (1454 entries) and significantly in the 'CustomerID' column (135,080 entries).
*   The dataset contained 5225 duplicate rows.
*   After cleaning, the DataFrame has 401,604 entries with no missing values or duplicates.
*   The 'Quantity' and 'UnitPrice' columns initially contained negative values and extremely high positive values, indicating the presence of outliers.
*   After filtering out negative/zero values and applying a 99th percentile cap, the 'Quantity' ranges from 1 to 120 and 'UnitPrice' ranges from \$0.001 to \$14.95.

### Insights or Next Steps

*   The presence of negative values in 'Quantity' and 'UnitPrice' suggests the need for careful consideration of return transactions or data entry errors in future analysis.
*   The significant number of missing 'CustomerID' values highlights a limitation in performing customer-specific analysis on the complete original dataset; dropping these rows was a necessary step for customer-centric tasks.
